File: src/main/java/sk/hudak/prco/ssl/PrcoCustomTrustManager.java
package sk.hudak.prco.ssl;

import lombok.extern.slf4j.Slf4j;

import javax.net.ssl.TrustManager;
import javax.net.ssl.TrustManagerFactory;
import javax.net.ssl.X509TrustManager;
import java.security.KeyStore;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.List;

@Slf4j
public class PrcoCustomTrustManager implements X509TrustManager {

    private X509TrustManager javaDefaultTrustManager;
    private X509Certificate serverCert;

    public PrcoCustomTrustManager() throws Exception {
        log.debug("inicializing");
        this.javaDefaultTrustManager = (X509TrustManager) getJavaDefaultTrustManager();
    }

    private TrustManager getJavaDefaultTrustManager() throws Exception {
        TrustManagerFactory tmg = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
        tmg.init((KeyStore) null);
        // beriem prvy
        return tmg.getTrustManagers()[0];
    }

    @Override
    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
        log.debug("checkClientTrusted()");
        log.debug("delegating to java default trust manager");
        javaDefaultTrustManager.checkClientTrusted(chain, authType);
    }

    @Override
    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
        this.serverCert = chain[0];
        String name = serverCert.getSubjectX500Principal().getName();
        log.debug("server cert subject name {}", name);
        // ak sa v subjecte certifikatu nachadza povoleny hostname, koncim
        // validaciu
        for (String allowedHostName : PrcoSSLContants.ALLOWED_HOSTNAME) {
            if (name.contains(allowedHostName)) {
                log.debug("ignoring server cert");
                return;
            }
        }
        log.debug("delegating checking to java default trust manager");
        this.javaDefaultTrustManager.checkServerTrusted(chain, authType);
    }

    @Override
    public X509Certificate[] getAcceptedIssuers() {
        log.debug("getAcceptedIssuers()");
        List<X509Certificate> issuers = new ArrayList<>();
        issuers.add(this.serverCert);
        for (X509Certificate javaDefaultIssuer : this.javaDefaultTrustManager.getAcceptedIssuers()) {
            issuers.add(javaDefaultIssuer);
        }
        return issuers.toArray(new X509Certificate[issuers.size()]);
    }

}

File: src/main/java/sk/hudak/prco/ssl/PrcoCustomTrustManager.kt
package sk.hudak.prco.ssl

import org.slf4j.LoggerFactory
import java.security.KeyStore
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.*
import javax.net.ssl.TrustManager
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager


class PrcoCustomTrustManager @Throws(Exception::class)
constructor() : X509TrustManager {

    companion object {
        val log = LoggerFactory.getLogger(PrcoCustomTrustManager::class.java)
    }

    private val javaDefaultTrustManager: X509TrustManager
    private var serverCert: X509Certificate? = null

    init {
        log.debug("initializing")
        this.javaDefaultTrustManager = getJavaDefaultTrustManager() as X509TrustManager
    }

    @Throws(Exception::class)
    private fun getJavaDefaultTrustManager(): TrustManager {
        val tmg = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
        tmg.init(null as KeyStore?)
        // beriem prvy
        return tmg.trustManagers[0]
    }

    @Throws(CertificateException::class)
    override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {
        log.debug("checkClientTrusted()")
        log.debug("delegating to java default trust manager")
        javaDefaultTrustManager.checkClientTrusted(chain, authType)
    }

    @Throws(CertificateException::class)
    override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
        this.serverCert = chain[0]
        val name = serverCert!!.subjectX500Principal.name
        log.debug("server cert subject name $name")
        // ak sa v subjecte certifikatu nachadza povoleny hostname, koncim
        // validaciu
        for (allowedHostName in PrcoSSLContants.ALLOWED_HOSTNAME) {
            if (name.contains(allowedHostName)) {
                log.trace("ignoring server cert")
                return
            }
        }
        log.debug("delegating checking to java default trust manager")
        this.javaDefaultTrustManager.checkServerTrusted(chain, authType)
    }

    override fun getAcceptedIssuers(): Array<X509Certificate> {
        log.trace("getAcceptedIssuers()")
        val issuers = ArrayList<X509Certificate>()
        issuers.add(serverCert!!)
        Collections.addAll(issuers, *this.javaDefaultTrustManager.acceptedIssuers)
        return issuers.toTypedArray()
    }

}